model/validation.go

```diff
@@ -233,59 +233,84 @@ func ValidateFQDN(fqdn string) error {
+// ValidateIstioService checks for validity of a service reference
+func ValidateIstioService(svc *proxyconfig.IstioService) (errs error) {
+	if svc.Name == "" {
+		errs = multierror.Append(errs, errors.New("name is mandatory for a service reference"))
+	} else if !IsDNS1123Label(svc.Name) {
+		errs = multierror.Append(errs, fmt.Errorf("name %q must be a valid label", svc.Name))
+	}
+
+	if svc.Namespace != "" && !IsDNS1123Label(svc.Namespace) {
+		errs = multierror.Append(errs, fmt.Errorf("namespace %q must be a valid label", svc.Namespace))
+	}
+
+	if svc.Domain != "" {
+		if err := ValidateFQDN(svc.Domain); err != nil {
 			errs = multierror.Append(errs, err)
 		}
 	}
 
-	if err := Tags(mc.SourceTags).Validate(); err != nil {
+	// TODO: handle service reference
+
+	if err := Labels(svc.Labels).Validate(); err != nil {
 		errs = multierror.Append(errs, err)
 	}
 
-	if mc.GetTcp() != nil {
-		if err := ValidateL4MatchAttributes(mc.GetTcp()); err != nil {
+	return
+}
+
+// ValidateMatchCondition validates a match condition
+func ValidateMatchCondition(mc *proxyconfig.MatchCondition) (errs error) {
+	if mc.Source != nil {
+		if err := ValidateIstioService(mc.Source); err != nil {
 			errs = multierror.Append(errs, err)
 		}
 	}
 
-	if mc.GetUdp() != nil {
-		if err := ValidateL4MatchAttributes(mc.GetUdp()); err != nil {
+	if mc.Tcp != nil {
+		if err := ValidateL4MatchAttributes(mc.Tcp); err != nil {
 			errs = multierror.Append(errs, err)
 		}
-		errs = multierror.Append(errs, fmt.Errorf("Istio does not support UDP protocol yet"))
 	}
 
-	for name, value := range mc.GetHttpHeaders() {
-		if err := ValidateHTTPHeaderName(name); err != nil {
-			errs = multierror.Append(errs, multierror.Prefix(err, fmt.Sprintf("header name %q invalid: ", name)))
-		}
-		if err := ValidateStringMatch(value); err != nil {
-			errs = multierror.Append(errs, multierror.Prefix(err, fmt.Sprintf("header %q value invalid: ", name)))
+	if mc.Udp != nil {
+		if err := ValidateL4MatchAttributes(mc.Udp); err != nil {
+			errs = multierror.Append(errs, err)
 		}
+		errs = multierror.Append(errs, fmt.Errorf("Istio does not support UDP protocol yet"))
+	}
 
-		// validate special `uri` header:
-		// absolute path must be non-empty (https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2)
-		if name == HeaderURI {
-			switch m := value.MatchType.(type) {
-			case *proxyconfig.StringMatch_Exact:
-				if m.Exact == "" {
-					errs = multierror.Append(errs, fmt.Errorf("exact header value for %q must be non-empty", HeaderURI))
-				}
-			case *proxyconfig.StringMatch_Prefix:
-				if m.Prefix == "" {
-					errs = multierror.Append(errs, fmt.Errorf("prefix header value for %q must be non-empty", HeaderURI))
-				}
-			case *proxyconfig.StringMatch_Regex:
-				if m.Regex == "" {
-					errs = multierror.Append(errs, fmt.Errorf("regex header value for %q must be non-empty", HeaderURI))
+	if mc.Request != nil {
+		for name, value := range mc.Request.Headers {
```

Comment:
Can you add a todo here to adopt the mixer style request attributes as first class fields instead of special header types ?

Reply:
request.headers is not using attributes. Are you suggesting adding request.* attributes in addition to headers?

Reply:
Yes. So that we can have things like user-agent , method, etc as first class fields. That said, it's not a priority and need to double check on the utility of that. Config doesn't differ much. In the end, I've convinced myself that this is of not much use to us.